src/dune_rules/coq/coq_lib.ml

```diff
@@ -244,32 +387,46 @@ module DB = struct
       let+ () =
         if allow_private_deps then Resolve.Memo.return ()
         else
-          match theory.package with
-          | Some _ -> Resolve.Memo.return ()
-          | None -> Error.private_deps_not_allowed ~loc theory_name
+          match theory with
+          | Dune { package = None; _ } ->
+            Error.private_deps_not_allowed ~loc theory_name
+          | Legacy _ | Dune _ -> Resolve.Memo.return ()
       in
       (loc, theory)
 
     let resolve_theories ~coq_lang_version ~allow_private_deps ~coq_db ~boot_id
-        theories =
+        ~loc ~use_stdlib ~is_boot theories =
+      (* If the user has not provided Coq as a dependency and (use_stdlib yes)
+         is set, we add it here. This will be used in extraction and dune coq
+         top. *)
+      let theories =
+        if
+          use_stdlib && (not is_boot)
+          && not
+             @@ List.exists theories ~f:(fun (_, name) ->
+                    Coq_lib_name.equal Coq_lib_name.stdlib name)
+        then (loc, Coq_lib_name.stdlib) :: theories
+        else theories
```

Comment:
Why are we doing this? The coq lib should already be in the list of theories by virtue of `maybe_add_boot` below.

Reply:
If we don't use the stdlib, then we don't want it included in theories. Take for example building the stdlib itself. I agree that there is some overlap of the boot detection behaviour and (stdlib yes). Not sure how to clean this up further though.

Reply:
So actually I can see why you would need this, in the sense that it seems we are not setting the boot library correctly when it is living in the parent.

Mmm, so it seems to me that Dune `main` may be buggy for the scenario where a library with its own scope is composed with Coq ?

Reply:
Well, the code in main seems to be working fine.

Reply:
And the code here should not need this because we are setting the parent correctly

Reply:
Well it is only buggy if we pass `-boot`. Since we do not we don't observe any difference.

Reply:
As this is the main blocker, if you push a commit removing this code, what breaks?

Reply:
I'm conjecturing nothing will happen, likely this code was needed before the last refactoring, but I think it should now be fine as we don't do the foward ref hack.

Reply:
Something did break last time I tried, I will try it again soon and report back.

Reply:
Thanks, let's indeed see how things break then.